src/go/wsl-helper/cmd/kubeconfig.go

```diff
@@ -42,80 +46,174 @@ var kubeconfigCmd = &cobra.Command{
-					return nil
-				}
-				return err
-			}
-		} else {
-			// No need to create if we want to remove it
-			target, err := os.Readlink(linkPath)
-			if err != nil {
-				if errors.Is(err, os.ErrNotExist) {
-					return nil
-				}
-				return err
-			}
-			if target == configPath {
-				err = os.Remove(linkPath)
-				if err != nil && !errors.Is(err, os.ErrNotExist) {
-					return err
-				}
-			}
+
+		cleanConfig := removeExistingRDConfig(rdCluster, &linuxConfig)
+
+		kubeConfig, err := updateClusterIP(winConfig, *cleanConfig, rdNetworking)
+
+		var finalKubeConfigFile *os.File
+		if err := os.MkdirAll(linuxConfigDir, 0o750); err != nil {
+			return err
+		}
+		finalKubeConfigFile, err = os.Create(filepath.Join(linuxConfigDir, "config"))
+		if err != nil {
+			return err
+		}
+		defer finalKubeConfigFile.Close()
+		err = os.MkdirAll(linuxConfigDir, 0o750)
+		if err != nil && !errors.Is(err, os.ErrExist) {
+			// The error already contains the full path, we can't do better.
+			return err
+		}
+		err = yaml.NewEncoder(finalKubeConfigFile).Encode(kubeConfig)
+		if err != nil {
+			return err
 		}
 		return nil
 	},
 }
 
+func readKubeConfig(configPath string) (kubeConfig, error) {
+	var config kubeConfig
+	configFile, err := os.Open(configPath)
+	if err != nil {
+		return config, err
+	}
+	defer configFile.Close()
+	err = yaml.NewDecoder(configFile).Decode(&config)
+	if err != nil {
+		return config, err
+	}
+
+	return config, nil
+}
+
+// updateClusterIP reads the kube config from windows side it also
+// modifies the cluster's server host to an appropriate address.
+// It then merges the config with an exisiting configuration on
+// users distro and returns the merged config.
+func updateClusterIP(winConfig, linuxConfig kubeConfig, rdNetworking bool) (kubeConfig, error) {
+	ip, err := getClusterIP()
+	if err != nil {
+		return winConfig, err
+	}
+
+	for clusterIdx, cluster := range winConfig.Clusters {
+		// Ignore any non rancher-desktop clusters
+		if winConfig.Clusters[clusterIdx].Name != rdCluster {
+			continue
+		}
+		server, err := url.Parse(cluster.Cluster.Server)
+		if err != nil {
+			// Ignore any clusters with invalid servers
+			continue
+		}
+		host := "gateway.rancher-desktop.internal"
+		if !rdNetworking {
+			host = ip.String()
+		}
+		if server.Port() != "" {
+			host = net.JoinHostPort(host, server.Port())
+		}
+		server.Host = host
+		winConfig.Clusters[clusterIdx].Cluster.Server = server.String()
+	}
+	return mergeKubeConfigs(winConfig, linuxConfig), nil
+}
+
+func removeExistingRDConfig(name string, config *kubeConfig) *kubeConfig {
+	// Remove clusters with the specified name
+	var filteredClusters []struct {
+		Cluster struct {
+			Server string
+			Extras map[string]interface{} `yaml:",inline"`
+		} `yaml:"cluster"`
+		Name   string                 `yaml:"name"`
+		Extras map[string]interface{} `yaml:",inline"`
+	}
+	for _, cluster := range config.Clusters {
```

Comment:
FWIW, I tested the thing now: https://go.dev/play/p/vZyZ77sT_yZ

So that's an easier way to do things, if you want.